File: dart/lib/src/vulgarity_preset.dart
import 'dart:convert';

import 'json_read.dart';
import 'model/vulgarity_term.dart';
import 'normalization/fold_table.g.dart';
import 'vulgarity_options.dart';

/// The preset schema this build reads.
const int kSupportedPresetSchema = 1;

/// The severity a preset entry takes when it states none.
///
/// A preset entry is added one at a time and by hand, so an entry that says
/// nothing sits in the middle of the range. A seed entry defaults to 1
/// instead. Both ports use these two numbers.
const int kPresetDefaultSeverity = 3;

/// The shape of a language code: 2 to 8 lower-case letters.
///
/// A preset only has to be well-formed here. Whether this build can serve the
/// code is a question for the builder, which may have a resolver that carries
/// packs this package never heard of.
final RegExp _languageCode = RegExp(r'^[a-z]{2,8}$');

/// A whole filter policy in one JSON document: the options, the extra terms,
/// the allowlist, and the terms to drop.
///
/// A seed document carries terms alone. A preset carries the policy too, so a
/// server can change how strict a client is without an app release.
///
/// Treat a preset from the network as untrusted. [VulgarityPreset.parse]
/// validates every field and throws [FormatException] with the offending field
/// named. It never partly applies a bad document.
///
/// ```dart
/// final response = await http.get(Uri.parse('https://example.com/policy.json'));
/// final filter = VulgarityFilter.fromPreset(response.body);
/// ```
class VulgarityPreset {
  const VulgarityPreset({
    this.name,
    this.description,
    this.languages = const <String>[],
    required this.options,
    this.entries = const <VulgarityTerm>[],
    this.allow = const <String>[],
    this.remove = const <String>[],
  });

  /// A short name for the policy.
  final String? name;

  /// What the policy is for.
  final String? description;

  /// The term lists to load, by language code.
  ///
  /// [VulgarityPreset.parse] checks the shape of each code and nothing more.
  /// Whether a code can be served is settled later, by
  /// [VulgarityFilterBuilder.addPreset] and the resolver it was given.
  final List<String> languages;

  /// The policy. A missing field in the document keeps its default.
  final VulgarityOptions options;

  /// Extra terms the preset adds.
  final List<VulgarityTerm> entries;

  /// Innocent words the filter must never flag.
  final List<String> allow;

  /// Terms to drop after the language packs load.
  final List<String> remove;

  /// Reads a preset document.
  ///
  /// Throws [FormatException] when the document is malformed, when it targets
  /// another fold profile, or when a field is out of range. Every field is
  /// type-checked rather than coerced, so `"sev": "3"` and `"name": 5` are
  /// both errors.
  factory VulgarityPreset.parse(String json) {
    final Object? parsed;
    try {
      parsed = jsonDecode(json);
    } on FormatException catch (error) {
      throw FormatException('The preset is not valid JSON. ${error.message}');
    }

    if (parsed is! Map<String, dynamic>) {
      throw const FormatException('A preset must be a JSON object.');
    }

    final int? schema = readOptionalInt(parsed, 'schema');
    if (schema != null && schema != kSupportedPresetSchema) {
      throw FormatException(
          'This build reads preset schema $kSupportedPresetSchema. '
          'The document states schema $schema.');
    }

    // The profile pins the fold table. A preset built against an older table
    // would match differently, so it fails here.
    final String? profile = readOptionalString(parsed, 'profile');
    if (profile != null && profile != kFoldProfile) {
      throw FormatException(
          "This build implements fold profile '$kFoldProfile'. "
          "The preset states '$profile'.");
    }

    final List<VulgarityTerm> entries = <VulgarityTerm>[];
    final Object? rawEntries = parsed['entries'];
    if (rawEntries != null) {
      if (rawEntries is! List) {
        throw FormatException("'entries' must be an array. "
            'The document states ${jsonTypeName(rawEntries)}.');
      }
      for (final Object? entry in rawEntries) {
        entries.add(_readEntry(entry));
      }
    }

    final Object? rawOptions = parsed['options'];
    if (rawOptions != null && rawOptions is! Map<String, dynamic>) {
      throw FormatException("'options' must be a JSON object. "
          'The document states ${jsonTypeName(rawOptions)}.');
    }

    return VulgarityPreset(
      name: readOptionalString(parsed, 'name'),
      description: readOptionalString(parsed, 'description'),
      languages: _readLanguages(parsed),
      options: rawOptions == null
          ? VulgarityOptions()
          : VulgarityOptions.fromJson(rawOptions as Map<String, dynamic>),
      entries: entries,
      allow: readStrings(parsed, 'allow'),
      remove: readStrings(parsed, 'remove'),
    );
  }

  /// Writes this preset as a JSON document.
  Map<String, dynamic> toJson() {
    return <String, dynamic>{
      'schema': kSupportedPresetSchema,
      'profile': kFoldProfile,
      if (name != null) 'name': name,
      if (description != null) 'description': description,
      'languages': languages,
      'options': options.toJson(),
      'entries': entries
          .map((VulgarityTerm t) => <String, dynamic>{
                't': t.text,
                'cat': t.categoryName,
                'sev': t.severity,
                if (t.requireBoundary) 'w': true,
              })
          .toList(),
      'allow': allow,
      'remove': remove,
    };
  }

  /// Reads and shape-checks the language codes.
  ///
  /// This checks the SHAPE only. It deliberately does not ask whether this
  /// build carries the code, because a caller may pass a resolver that serves
  /// codes no bundled list covers.
  static List<String> _readLanguages(Map<String, dynamic> root) {
    final List<String> codes = readStrings(root, 'languages');
    for (final String code in codes) {
      if (!_languageCode.hasMatch(code)) {
        throw FormatException("'languages' names '$code', which is not a "
            'language code. A code is 2 to 8 lower-case letters.');
      }
    }
    return codes;
  }

  static VulgarityTerm _readEntry(Object? entry) {
    if (entry is! Map<String, dynamic>) {
      throw const FormatException(
          "Every entry in 'entries' must be a JSON object.");
    }

    final Object? term = entry['t'];
    if (term is! String || term.isEmpty) {
      throw const FormatException(
          "Every entry in 'entries' must hold a non-empty 't' term.");
    }

    // A term category the build does not know maps to other. That keeps an
    // older client working when a server adds a category. A category that is
    // not a string is a different thing: that is a malformed document.
    final String category = readOptionalString(entry, 'cat') ?? 'other';

    final int sev = readOptionalInt(entry, 'sev') ?? kPresetDefaultSeverity;
    if (sev < 1 || sev > 5) {
      throw FormatException(
          "Severity must be 1 to 5. Term '$term' states $sev.");
    }

    return VulgarityTerm(
      term,
      category,
      sev,
      readOptionalBool(entry, 'w') ?? false,
    );
  }

  @override
  String toString() => 'preset ${name ?? '(unnamed)'}: '
      '${languages.length} languages, ${entries.length} added, '
      '${remove.length} removed';
}

File: dotnet/src/Vulgarity/VulgarityPreset.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using System.Text.Json;
using Vulgarity.Normalization;

namespace Vulgarity
{
    /// <summary>
    /// A whole filter policy in one JSON document: the options, the extra
    /// terms, the allowlist, and the terms to drop.
    /// </summary>
    /// <remarks>
    /// <para>
    /// A seed document carries terms alone. A preset carries the policy too, so
    /// a server can change how strict a client is without an app release.
    /// </para>
    /// <para>
    /// Treat a preset from the network as untrusted. <see cref="Parse"/>
    /// validates every field and throws <see cref="FormatException"/> with the
    /// offending field named. It never partly applies a bad document.
    /// </para>
    /// <example>
    /// <code>
    /// string json = await http.GetStringAsync("https://example.com/policy.json");
    /// var filter = VulgarityFilter.FromPreset(json);
    /// </code>
    /// </example>
    /// </remarks>
    public sealed class VulgarityPreset
    {
        /// <summary>The preset schema this build reads.</summary>
        public const int SupportedSchema = 1;

        internal VulgarityPreset(
            string name,
            string description,
            IReadOnlyList<string> languages,
            VulgarityOptions options,
            IReadOnlyList<VulgarityTerm> entries,
            IReadOnlyList<string> allow,
            IReadOnlyList<string> remove)
        {
            Name = name;
            Description = description;
            Languages = languages;
            Options = options;
            Entries = entries;
            Allow = allow;
            Remove = remove;
        }

        /// <summary>A short name for the policy. It can be null.</summary>
        public string Name { get; private set; }

        /// <summary>What the policy is for. It can be null.</summary>
        public string Description { get; private set; }

        /// <summary>The bundled term lists to load, by language code.</summary>
        public IReadOnlyList<string> Languages { get; private set; }

        /// <summary>The policy. It is never null; a missing field keeps its default.</summary>
        public VulgarityOptions Options { get; private set; }

        /// <summary>Extra terms the preset adds.</summary>
        public IReadOnlyList<VulgarityTerm> Entries { get; private set; }

        /// <summary>Innocent words the filter must never flag.</summary>
        public IReadOnlyList<string> Allow { get; private set; }

        /// <summary>Terms to drop after the language packs load.</summary>
        public IReadOnlyList<string> Remove { get; private set; }

        /// <summary>Builds a preset that loads the bundled English list and nothing else.</summary>
        public static VulgarityPreset Default()
        {
            return new VulgarityPreset(
                "default",
                "The bundled English term list with default options.",
                new[] { VulgarityLanguages.Default },
                new VulgarityOptions(),
                new VulgarityTerm[0],
                new string[0],
                new string[0]);
        }

        /// <summary>Reads a preset document.</summary>
        /// <exception cref="FormatException">The document is malformed, or a field is out of range.</exception>
        public static VulgarityPreset Parse(string json)
        {
            if (json == null)
            {
                throw new ArgumentNullException("json");
            }

            JsonDocument document;
            try
            {
                document = JsonDocument.Parse(json);
            }
            catch (JsonException error)
            {
                throw new FormatException("The preset is not valid JSON. " + error.Message, error);
            }

            using (document)
            {
                JsonElement root = document.RootElement;
                if (root.ValueKind != JsonValueKind.Object)
                {
                    throw new FormatException("A preset must be a JSON object.");
                }

                JsonElement value;

                if (root.TryGetProperty("schema", out value) && value.ValueKind != JsonValueKind.Null)
                {
                    int schema = value.GetInt32();
                    if (schema != SupportedSchema)
                    {
                        throw new FormatException(
                            "This build reads preset schema " + SupportedSchema +
                            ". The document states schema " + schema + ".");
                    }
                }

                // The profile pins the fold table. A preset built against an
                // older table would match differently, so it fails here.
                if (root.TryGetProperty("profile", out value) && value.ValueKind != JsonValueKind.Null)
                {
                    string profile = value.GetString();
                    if (profile != FoldTableData.Profile)
                    {
                        throw new FormatException(
                            "This build implements fold profile '" + FoldTableData.Profile +
                            "'. The preset states '" + profile + "'.");
                    }
                }

                string name = root.TryGetProperty("name", out value) && value.ValueKind == JsonValueKind.String
                    ? value.GetString()
                    : null;
                string description =
                    root.TryGetProperty("description", out value) && value.ValueKind == JsonValueKind.String
                        ? value.GetString()
                        : null;

                List<string> languages = ReadStrings(root, "languages");
                foreach (string code in languages)
                {
                    if (!VulgarityLanguages.Has(code))
                    {
                        throw new FormatException(
                            "'languages' names '" + code + "', which this build does not carry. Available: "
                            + string.Join(", ", new List<string>(VulgarityLanguages.Available).ToArray()) + ".");
                    }
                }

                VulgarityOptions options =
                    root.TryGetProperty("options", out value) && value.ValueKind != JsonValueKind.Null
                        ? VulgarityOptions.FromElement(value)
                        : new VulgarityOptions();

                List<VulgarityTerm> entries = new List<VulgarityTerm>();
                if (root.TryGetProperty("entries", out value) && value.ValueKind != JsonValueKind.Null)
                {
                    if (value.ValueKind != JsonValueKind.Array)
                    {
                        throw new FormatException("'entries' must be an array.");
                    }

                    foreach (JsonElement entry in value.EnumerateArray())
                    {
                        entries.Add(ReadEntry(entry));
                    }
                }

                return new VulgarityPreset(
                    name,
                    description,
                    languages,
                    options,
                    entries,
                    ReadStrings(root, "allow"),
                    ReadStrings(root, "remove"));
            }
        }

        /// <summary>Writes this preset as a JSON document.</summary>
        public string ToJson()
        {
            using (MemoryStream stream = new MemoryStream())
            {
                using (Utf8JsonWriter writer = new Utf8JsonWriter(
                    stream, new JsonWriterOptions { Indented = true }))
                {
                    writer.WriteStartObject();
                    writer.WriteNumber("schema", SupportedSchema);
                    writer.WriteString("profile", FoldTableData.Profile);

                    if (Name != null)
                    {
                        writer.WriteString("name", Name);
                    }

                    if (Description != null)
                    {
                        writer.WriteString("description", Description);
                    }

                    writer.WriteStartArray("languages");
                    foreach (string code in Languages)
                    {
                        writer.WriteStringValue(code);
                    }

                    writer.WriteEndArray();

                    writer.WriteStartObject("options");
                    Options.WriteTo(writer);
                    writer.WriteEndObject();

                    writer.WriteStartArray("entries");
                    foreach (VulgarityTerm term in Entries)
                    {
                        writer.WriteStartObject();
                        writer.WriteString("t", term.Text);
                        writer.WriteString("cat", term.CategoryName);
                        writer.WriteNumber("sev", term.Severity);
                        if (term.RequireBoundary)
                        {
                            writer.WriteBoolean("w", true);
                        }

                        writer.WriteEndObject();
                    }

                    writer.WriteEndArray();

                    WriteStrings(writer, "allow", Allow);
                    WriteStrings(writer, "remove", Remove);
                    writer.WriteEndObject();
                }

                return Encoding.UTF8.GetString(stream.ToArray());
            }
        }

        private static void WriteStrings(Utf8JsonWriter writer, string name, IReadOnlyList<string> values)
        {
            writer.WriteStartArray(name);
            foreach (string value in values)
            {
                writer.WriteStringValue(value);
            }

            writer.WriteEndArray();
        }

        private static List<string> ReadStrings(JsonElement root, string name)
        {
            List<string> result = new List<string>();
            JsonElement value;
            if (!root.TryGetProperty(name, out value) || value.ValueKind == JsonValueKind.Null)
            {
                return result;
            }

            if (value.ValueKind != JsonValueKind.Array)
            {
                throw new FormatException("'" + name + "' must be an array of strings.");
            }

            foreach (JsonElement item in value.EnumerateArray())
            {
                if (item.ValueKind != JsonValueKind.String)
                {
                    throw new FormatException("'" + name + "' must hold strings only.");
                }

                string text = item.GetString();
                if (!string.IsNullOrEmpty(text))
                {
                    result.Add(text);
                }
            }

            return result;
        }

        private static VulgarityTerm ReadEntry(JsonElement entry)
        {
            if (entry.ValueKind != JsonValueKind.Object)
            {
                throw new FormatException("Every entry in 'entries' must be a JSON object.");
            }

            JsonElement value;
            if (!entry.TryGetProperty("t", out value) || value.ValueKind != JsonValueKind.String)
            {
                throw new FormatException("Every entry in 'entries' must hold a 't' term.");
            }

            string term = value.GetString();
            if (string.IsNullOrEmpty(term))
            {
                throw new FormatException("A term in 'entries' must not be empty.");
            }

            // A term category the build does not know maps to Other. That keeps
            // an older client working when a server adds a category.
            string category = entry.TryGetProperty("cat", out value) && value.ValueKind == JsonValueKind.String
                ? value.GetString()
                : "other";

            int severity = 3;
            if (entry.TryGetProperty("sev", out value) && value.ValueKind != JsonValueKind.Null)
            {
                severity = value.GetInt32();
                if (severity < 1 || severity > 5)
                {
                    throw new FormatException(
                        "Severity must be 1 to 5. Term '" + term + "' states " + severity + ".");
                }
            }

            bool requireBoundary = entry.TryGetProperty("w", out value) && value.ValueKind == JsonValueKind.True;
            return new VulgarityTerm(term, category, severity, requireBoundary);
        }

        public override string ToString()
        {
            return "preset " + (Name ?? "(unnamed)")
                + ": " + Languages.Count + " languages, " + Entries.Count + " added, "
                + Remove.Count + " removed";
        }
    }
}
